### ISSUE TITLE
Does it has something wrong in reconstruct loss code?

### ISSUE THREAD

**Issue by AlexHex7** [timomernick/pytorch-capsule#2]
```
        masked = Variable(torch.zeros(input.size())).cuda()
        masked[:,v_max_index] = input[:,v_max_index]
```
The version of python I use is 3.5.
In my opinion, the v_max_index (batchsize, 1) means for each sample of batch, there is a max length vector among 16 vectors. So for the sample 0, there should be active vector among 16 vectors, and other 15 vectors are all 0.
But for

`masked[:,v_max_index] = input[:,v_max_index]`

it means each sample, there are v_max_index vector will be assign which includes duplicate assignments.

And the paper says
> During training, we mask out all but the activity vector of the correct digit capsule. 

The figure 2 in the paper show that for each sample, there are 15 vectors among 16 vectors will be masked.

**Comment by timomernick** [comment]
Thanks for the comment.  I think it's not the most efficient, but it does work as intended.  It's easiest to see if you set batch_size to 1, and reduce output_unit_size to to something like 4.  Then print out v_max_index and masked.  For example, when v_max_index is '1', you get this, with all but the first 4-vector in 'masked' zeroed out:

```v_max_index=1

[[[[  0.00000000e+00]
   [  0.00000000e+00]
   [  0.00000000e+00]
   [  0.00000000e+00]]

  [[  6.14560843e-02]
   [ -5.99865057e-02]
   [  2.03099262e-06]
   [ -5.79577461e-02]]

  [[  0.00000000e+00]
   [  0.00000000e+00]
   [  0.00000000e+00]
   [  0.00000000e+00]]

  [[  0.00000000e+00]
   [  0.00000000e+00]
   [  0.00000000e+00]
   [  0.00000000e+00]]

  [[  0.00000000e+00]
   [  0.00000000e+00]
   [  0.00000000e+00]
   [  0.00000000e+00]]

  [[  0.00000000e+00]
   [  0.00000000e+00]
   [  0.00000000e+00]
   [  0.00000000e+00]]

  [[  0.00000000e+00]
   [  0.00000000e+00]
   [  0.00000000e+00]
   [  0.00000000e+00]]

  [[  0.00000000e+00]
   [  0.00000000e+00]
   [  0.00000000e+00]
   [  0.00000000e+00]]

  [[  0.00000000e+00]
   [  0.00000000e+00]
   [  0.00000000e+00]
   [  0.00000000e+00]]

  [[  0.00000000e+00]
   [  0.00000000e+00]
   [  0.00000000e+00]
   [  0.00000000e+00]]]]```

**Comment by timomernick** [comment]
Oh I just found what you mean.  It's duplicating the assignments over the batch dimension.  I'll fix it, thanks for pointing this out!

**Comment by timomernick** [comment]
To verify, here is what the masked capsule activity looks like after my fix.  Only one capsule is nonzero for each batch sample.  Here there is a batch size of 2.  The first sample has capsule 0 active, and the second sample has capsule 2 active.

```('all_masked', array([[[[ -7.67778372e-04],
         [ -3.17241793e-04],
         [ -7.09305750e-05],
         [  2.88858573e-05]],

        [[  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00]],

        [[  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00]],

        [[  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00]],

        [[  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00]],

        [[  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00]],

        [[  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00]],

        [[  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00]],

        [[  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00]],

        [[  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00]]],


       [[[  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00]],

        [[  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00]],

        [[ -3.76116368e-04],
         [  2.83236353e-04],
         [  1.13728966e-04],
         [ -8.82451714e-05]],

        [[  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00]],

        [[  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00]],

        [[  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00]],

        [[  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00]],

        [[  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00]],

        [[  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00]],

        [[  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00],
         [  0.00000000e+00]]]], dtype=float32))
```
